hips/hip-00XX.md

```diff
@@ -0,0 +1,136 @@
+
+https://kubernetes.io/blog/2022/10/20/advanced-server-side-apply/
+
+Helm supporting SSA will allow users to realize the benefits of SSA. And longer term goals (out of scope for this HIP) might be for Helm to utilize SSA by default ([similar to kubectl](https://github.com/kubernetes/enhancements/tree/master/keps/sig-cli/3805-ssa-default)). Finally later as a code mantainence benefit, eventually drop support for the strategic-merge patch client-side apply process/code.
+
+## Rationale
+
+<!-- Describe why particular design decisions were made. -->
+
+Due to the backwards compatibility concerns documented below, the decision was made to provide opt-in support only. And even when a chart is deployed with SSA enabled, Helm should mantain full compatibility with reverting back to using CSA for upgrades/rollbacks. Additionally it was desirable to mimic kubectl's cli interface/flags. As a way for Helm to be consistent with an existing kubectl functionality.
+
+## Specification
+
+<!-- Describe the syntax and semantics of any new feature. -->
+
+### General
+
+Helm will add a new flag to the install, upgrade and rollback sub-commands, modelled after the like `kubectl` flag. And add a corresponding field to the respective install/upgrade/rollback SDK API objects:
+`--server-side=false|true|auto`
+
+Helm will default to `false` for installs, retaining its existing behaviour. When `true`, Helm will submit resources to the Kubernetes API server with SSA enabled. When `auto`, Helm will default to CSA on installs (`false`), and for upgrades and rollbacks enable SSA based on the whether SSA was utilized for the prior operation (`auto`). The usage of SSA for a given release will be stored in the release's manifest.
+
+If a user deploys a chart with a new version of Helm with SSA enabled, then upgrades the chart with an older version of Helm. It is expected that the older Helm will be able to upgrade the chart with CSA. And the new manifest object will omit the indication of SSA usage from the releases manifest (and so futher upgrades/rollbacks will continue to use CSA by default).
+
+In a future version of Helm 3 cli, it might be desirable to omit a warning if a user is not using SSA (in order to prompt the user to enabling SSA).
+
+#### Custom resources
+
+Helm will utilize SSA for custom resources, which previously it replaced (TODO: reference). As SSA allows custom resources to be correctly merged based on metadata within the CRD. Or the API server will use the default approach <https://kubernetes.io/docs/reference/using-api/server-side-apply/#custom-resources>.
+
+#### Conflicts and forcing
+
+It is possible when Helm upgrades a chart, that there will be a field conflict with another field manager. In this case, Helm should report the error to the user.
```

Comment:
TODO: verify we can abort atomically across multiple objects (Ian Z points this out)